Changes in module 'nf-core/picard/crosscheckfingerprints'
--- modules/nf-core/picard/crosscheckfingerprints/main.nf
+++ modules/nf-core/picard/crosscheckfingerprints/main.nf
@@ -8,9 +8,8 @@
         'biocontainers/picard:3.1.1--hdfd78af_0' }"
 
     input:
-    tuple val(meta), path(input1)
-    path input2
-    path haplotype_map
+    tuple val(meta),  path(input1), path(input2), path(haplotype_map)
+    tuple val(meta2), path(fasta)
 
     output:
     tuple val(meta), path("*.crosscheck_metrics.txt"), emit: crosscheck_metrics
@@ -23,8 +22,9 @@
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
 
-    def input1_string = input1.join(" --INPUT ")
-    def input2_string = input2 ? "--SECOND_INPUT " + input2.join(" --SECOND_INPUT ") : ""
+    def input1_cmd = input1.collect{"--INPUT $it"}.join(' ')
+    def input2_cmd = input2.collect{"--SECOND_INPUT $it"}.join(' ')
+    def reference_cmd = fasta ? "--REFERENCE_SEQUENCE $fasta" : ""
 
     def avail_mem = 3072
     if (!task.memory) {
@@ -36,16 +36,29 @@
     picard \\
         -Xmx${avail_mem}M \\
         CrosscheckFingerprints \\
-        $args \\
+        ${input1_cmd} \\
+        ${input2_cmd} \\
+        ${reference_cmd} \\
+        --HAPLOTYPE_MAP ${haplotype_map} \\
+        --OUTPUT ${prefix}.crosscheck_metrics.txt \\
         --NUM_THREADS ${task.cpus} \\
-        --INPUT $input1_string \\
-        $input2_string \\
-        --HAPLOTYPE_MAP ${haplotype_map} \\
-        --OUTPUT ${prefix}.crosscheck_metrics.txt
+        $args
+
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         picard: \$( picard CrosscheckFingerprints --version 2>&1 | grep -o 'Version:.*' | cut -f2- -d: )
     END_VERSIONS
     """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}.crosscheck_metrics.txt
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        picard: \$(echo \$(picard CollectHsMetrics --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d:)
+    END_VERSIONS
+    """
 }

************************************************************
